jerry-ext/jerry-module/jerry-module.c

```diff
@@ -0,0 +1,209 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jmem.h"
+#include "jerryscript.h"
+#include "jerry-module.h"
+
+static void
+jerryx_module_link_insert (jerryx_module_header_t *link,
+                           jerryx_module_header_t **root)
+{
+  if (*root)
+  {
+    link->next = (*root);
+  }
+  (*root) = link;
+} /* jerryx_module_link_insert */
+
+static jerryx_module_header_t *
+jerryx_module_link_find (jerryx_module_header_t *root,
+                         const jerry_char_t *name,
+                         jerryx_module_header_t **previous)
+{
+  for (;root && strcmp ((char *) name, ((char *) root->name)); root = root->next)
+  {
+    if (previous)
+    {
+      (*previous) = root;
+    }
+  }
+
+  return root;
+} /* jerryx_module_link_find */
+
+static jerryx_module_header_t *
+jerryx_module_link_remove (const jerry_char_t *name,
+                           jerryx_module_header_t **root)
+{
+  jerryx_module_header_t *previous = NULL;
+  jerryx_module_header_t *return_value = jerryx_module_link_find ((*root), name, &previous);
+
+  if (return_value)
+  {
+    if (previous)
+    {
+      previous->next = return_value->next;
+    }
+    else
+    {
+      (*root) = return_value->next;
+    }
+    JERRYX_MODULE_HEADER_UNLINK (return_value);
+  }
+
+  return return_value;
+} /* jerryx_module_link_remove */
+
+static void
+jerryx_module_link_free (jerryx_module_header_t *root,
+                         void (*free_link) (jerryx_module_header_t *link))
+{
+  jerryx_module_header_t *next;
+  while (root)
+  {
+    next = root->next;
+    free_link (root);
+    root = next;
+  }
+} /* jerryx_module_link_free */
+
+typedef struct
+{
+  jerryx_module_header_t link;
+  jerry_value_t result;
+} jerryx_module_instance_t;
+
+static void
+jerryx_module_free_instance (jerryx_module_header_t *link)
+{
+  jerryx_module_instance_t *instance = (jerryx_module_instance_t *) link;
+  jerry_release_value (instance->result);
```

Comment:
... but what to pass it? The module has no place to store per-context info except in jerry-context, but that already has an API for deinit. I don't want to duplicate that API.